src/VatRates/TaxesEuropeDatabaseClient.php

```diff
@@ -0,0 +1,93 @@
+<?php declare(strict_types = 1);
+
+namespace SandwaveIo\Vat\VatRates;
+
+use DateTime;
+use SoapClient;
+use SoapFault;
+
+final class TaxesEuropeDatabaseClient implements ResolvesVatRates
+{
+    const WSDL = 'https://ec.europa.eu/taxation_customs/tedb/ws/VatRetrievalService.wsdl';
+    const SOAP_CONFIG = [
+        'version' => SOAP_1_1,
+        'cache_wsdl' => WSDL_CACHE_MEMORY,
+    ];
+    /* @see https://ec.europa.eu/taxation_customs/tedb/ws/VatRetrievalServiceType.xsd */
+    const RATE_TYPE_STANDARD = 'STANDARD';
+    const RATE_VALUE_TYPE_DEFAULT = 'DEFAULT';
+
+    private SoapClient $client;
+
+    public function __construct(?SoapClient $client = null)
+    {
+        $this->client = $client ?? new SoapClient(self::WSDL, self::SOAP_CONFIG);
+    }
+
+    /**
+     * Retrieve the default VAT rate for a given country. If the countries VAT rate cannot be resolved, null is returned.
+     * This also happens if the country does not exist in the European Tax Database.
+     *
+     * @param string        $countryCode ISO country code
+     * @param DateTime|null $dateTime    Date of checking, defaults to today.
+     *
+     * @return float|null
+     */
+    public function getDefaultVatRateForCountry(string $countryCode, ?DateTime $dateTime = null): ?float
+    {
+        $date = $dateTime ?? new DateTime();
+        $response = $this->call('retrieveVatRates', [
+            'retrieveVatRatesRequest' => [
+                'memberStates' => [$countryCode],
+                'situationOn' => $date->format('Y-m-d'),
+            ],
+        ]);
+
+        if (is_null($response)) {
+            return null;
+        }
+
+        return $this->parseRetrieveVatRateResponse($response, $countryCode, self::RATE_TYPE_STANDARD, self::RATE_VALUE_TYPE_DEFAULT);
+    }
+
+    private function parseRetrieveVatRateResponse(object $response, string $countryCode, string $rateType, string $rateValueType): ?float
+    {
+        if (! isset($response->vatRateResults)) {
+            return null;
+        }
+        foreach ($response->vatRateResults as $vatRateResult) {
```

Comment:
Are we sure this cant be for example a string at this point? Without reading the api spec this looks a bit agressive at first sight. The others checks in this function look allot more defensive.

Reply:
This is right, but the response is hydrated by the Soap library based on the given WSDL. So at this point we know this adheres to the spec.